core/ecschema-metadata/src/UnitConversion/UnitConverter.ts

```diff
@@ -51,18 +48,19 @@ export class UnitConverter {
       });
     }
 
-    const from = await this._uGraph.resolveUnit(fromSchemaItemName, fromSchema);
-    const to = await this._uGraph.resolveUnit(toSchemaItemName, toSchema);
+    const unitGraph = new UnitGraph(this._context);
```

Comment:
On the surface the UnitGraph has checks to ensure adding the same unit twice doesn't break unit conversion.  What is actually happening that causes this to break.

Any performance benefit from reusing the UnitGraph probably isn't a big performance issue because the caller should be caching UnitConversions returned.  But recreating the UnitGraph on each call makes a UnitConverter object somewhat pointless.  

@rschili @hl662

Reply:
Looks like when calling addUnit in concurrent setups, the addUnit leaves early when it finds the unit in it's graph already (UnitGraph.hasNode). It may occur though, that at this point, the graph is not fully build yet (aka promise that adds the unit and it subcomponents is still running), so the method was left early. Now, the subsequent reduce calls in the UnitConverter produce an incomplete map which makes the base unit check fail.

I pushed a fix, where we cache incomplete addUnit operations and if a unit is added again, it won't leave the method before the adding-in-progress-promise has been resolved.